File: LabVIEW/fit_lin.py
#!/usr/bin/python3
import numpy as np
#import pandas as pd
import sys

print('Begin the matching.\n')

path_out = sys.argv[1]+'/'
Imin = int(sys.argv[2])
Imax = int(sys.argv[3])


for k in range(Imin, Imax + 1):
	with open(path_out + 'CH2_' + str(k) + '.dat', 'r') as header_file:
		header_lines = [header_file.readline().strip(), header_file.readline().strip()]
	
	data1 = np.loadtxt(path_out + 'CH2_' + str(k) + '.dat', skiprows=2)
	x1 = data1[:, 0]
	y1 = data1[:, 1]


	data2 = np.loadtxt(path_out + 'CH1_' + str(k) + '.dat', skiprows=2)
	x2 = data2[:, 0]
	y2 = data2[:, 1]


	m = len(x1)
	n = len(x2)
	y_out = []

	print('File %i#'%k)
	for i in range(0, n):
		d = []
		for j in range(0, m):
			d.append(abs(x1[j] - x2[i]))

		min_index, min_value = min(enumerate(d), key=lambda x: x[1])
		if x2[i] - x1[min_index] > 0:
			j2 = min_index
			j1 = j2 - 1
		if x2[i] - x1[min_index] < 0:
			j1 = min_index
			j2 = j1 + 1
		else:
			j2 = min_index
			j1 = j2
		if j1 == j2:
			y_out.append(y1[j1])
		else:
			y_out.append(y1[j2] + (x2[i] - x1[j2])*((y1[j2] - y1[j1])/(x1[j2] - x1[j1])))

	with open(path_out + 'CH2_' + str(k) + '.dat', 'w+') as file_out:
		file_out.write(header_lines[0] + '\n')
		file_out.write(header_lines[1] + '\n')
		for i in range(0, n):
			file_out.write('%.6f\t%.6f\n'%(x2[i], y_out[i]))
print('Matching completed. Click "check" to control the number of lines.')

File: C++/fit_lin.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

int main(int argc, char *argv[]) {
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <output_path> <Imin> <Imax>\n";
        return 1;
    }

    std::cout << "Begin the matching.\n";

    std::string path_out = argv[1];
    if (path_out.back() != '/') {
        path_out += "/";
    }
    int Imin = std::stoi(argv[2]);
    int Imax = std::stoi(argv[3]);

    for (int k = Imin; k <= Imax; ++k) {
        std::ifstream header_file(path_out + "CH2_" + std::to_string(k) + ".dat");
        std::string header_lines[2];
        if (header_file.is_open()) {
            std::getline(header_file, header_lines[0]);
            std::getline(header_file, header_lines[1]);
            header_file.close();
        } else {
            std::cerr << "Unable to open file: " << path_out + "CH2_" + std::to_string(k) + ".dat\n";
            continue;
        }

        std::ifstream data_file1(path_out + "CH2_" + std::to_string(k) + ".dat");
        std::ifstream data_file2(path_out + "CH1_" + std::to_string(k) + ".dat");
        std::vector<double> x1, y1, x2, y2;
        if (data_file1.is_open() && data_file2.is_open()) {
            double x, y;
            while (data_file1 >> x >> y) {
                x1.push_back(x);
                y1.push_back(y);
            }
            while (data_file2 >> x >> y) {
                x2.push_back(x);
                y2.push_back(y);
            }
            data_file1.close();
            data_file2.close();
        } else {
            std::cerr << "Unable to open data file CH2_" << k << ".dat or CH1_" << k << ".dat\n";
            continue;
        }

        int m = x1.size();
        int n = x2.size();
        std::vector<double> y_out(n);

        std::cout << "File " << k << "#\n";
        for (int i = 0; i < n; ++i) {
            std::vector<double> d(m);
            for (int j = 0; j < m; ++j) {
                d[j] = std::abs(x1[j] - x2[i]);
            }

            auto min_index = std::min_element(d.begin(), d.end()) - d.begin();
            int j1, j2;
            if (x2[i] - x1[min_index] > 0) {
                j2 = min_index;
                j1 = j2 - 1;
            } else if (x2[i] - x1[min_index] < 0) {
                j1 = min_index;
                j2 = j1 + 1;
            } else {
                j2 = min_index;
                j1 = j2;
            }

            if (j1 == j2) {
                y_out[i] = y1[j1];
            } else {
                y_out[i] = y1[j2] + (x2[i] - x1[j2]) * ((y1[j2] - y1[j1]) / (x1[j2] - x1[j1]));
            }
        }

        std::ofstream file_out(path_out + "CH2_" + std::to_string(k) + ".dat");
        if (file_out.is_open()) {
            file_out << header_lines[0] << "\n";
            file_out << header_lines[1] << "\n";
            for (int i = 0; i < n; ++i) {
                file_out << std::fixed << std::setprecision(6) << x2[i] << "\t" << y_out[i] << "\n";
            }
            file_out.close();
        } else {
            std::cerr << "Unable to create output file: " << path_out + "CH2_" + std::to_string(k) + ".dat\n";
            continue;
        }
    }

    std::cout << "Matching completed. Click \"check\" to control the number of lines.\n";
    return 0;
}
